tests/test_model_upload.py

```diff
@@ -1,8 +1,122 @@
+            self.send_response(200)
+            self.send_header("Content-type", "application/json")
+            self.end_headers()
+            response = {"status": "success", "message": "Model Instance created successfully"}
+            self.wfile.write(json.dumps(response).encode("utf-8"))
+        elif self.path == "/api/v1/blobs/upload":
+            self.send_response(200)
+            self.send_header("Content-type", "application/json")
+            self.end_headers()
+            self.wfile.write(
+                json.dumps(
+                    {
+                        "token": "dummy",
+                        "createUrl": "http://localhost:7778",
+                        "status": "success",
+                        "message": "Here is your token and Url",
+                    }
+                ).encode("utf-8")
+            )
+        else:
+            self.send_response(200)
+            self.send_header("Content-Type", "application/json")
+            self.end_headers()
+            error_message = json.dumps({"error": f"bad: {self.path}"})
+            self.wfile.write(bytes(error_message, "utf-8"))
+
+
+class GcsAPIHandler(BaseHTTPRequestHandler):
+    def do_HEAD(self):  # noqa: N802
+        self.send_response(200)
+
+    def do_PUT(self):  # noqa: N802
+        self.send_response(200)
+        self.send_header("Content-type", "application/json")
+        self.end_headers()
+        self.wfile.write(json.dumps({"status": "success", "message": "File uploaded"}).encode("utf-8"))
+
 
 class TestModelUpload(BaseTestCase):
-    def test_model_upload(self):
-        with self.assertRaises(NotImplementedError):
-            kagglehub.model_upload()
+    def test_model_upload_with_invalid_handle(self):
+        with create_test_http_server(KaggleAPIHandler):
+            try:
+                with TemporaryDirectory() as temp_dir:
+                    test_filepath = Path(temp_dir) / "temp_test_file"
+                    test_filepath.touch()  # Create a temporary file in the temporary directory
+                    model_upload("invalid/invalid/invalid", temp_dir, "Apache 2.0", "model_type")
+                self.fail("Expected an exception")
+            except Exception as e:
+                logger.info(f"Caught an exception: {type(e).__name__}")
+
+    def test_model_upload_instance_with_valid_handle(self):
+        # execution path: get_model -> create_model -> get_instance -> create_version
+        with create_test_http_server(KaggleAPIHandler):
+            with create_test_http_server(GcsAPIHandler, "http://localhost:7778"):
+                try:
+                    with TemporaryDirectory() as temp_dir:
+                        test_filepath = Path(temp_dir) / "temp_test_file"
+                        test_filepath.touch()  # Create a temporary file in the temporary directory
+                        model_upload("valid/valid/valid/1", temp_dir, "Apache 2.0", "model_type")
```

Comment:
I will add client-side validation to the model handle soon, this would fail given the variation slug is too short. Use `metaresearch/new-model/pyTorch/new-variation`